Lua redis.log implementation

upcFrost opened the pull request:
Added a very simple `redis.log` Lua call support, wired to the standard python `logging` lib.

Issue #265

bmerry commented on fakeredis/_server.py:
```diff
@@ -2312,6 +2322,12 @@ def _lua_redis_pcall(self, lua_runtime, expected_globals, op, *args):
+        self._check_for_lua_globals(lua_runtime, expected_globals)
+        if len(args) == 1:
+            msg = args[0]
+            LOGGER.log(lvl or logging.INFO, msg.decode('utf-8') if isinstance(msg, bytes) else msg)
```
What does real redis do if the message is not a string (e.g. is an integer)?

bmerry commented on test_fakeredis.py:
```diff
@@ -4160,6 +4160,24 @@ def test_script(self):
         self.assertEqual(result, b'42')
 
+    def test_lua_log(self):
+        if not isinstance(self.redis, fakeredis.FakeRedis) and not isinstance(self.redis, fakeredis.FakeStrictRedis):
```
```suggestion
        if not isinstance(self.redis, (fakeredis.FakeRedis, fakeredis.FakeStrictRedis)):
```

upcFrost replied:
Fixed

bmerry commented on test_fakeredis.py:
```diff
@@ -4160,6 +4160,24 @@ def test_script(self):
+                                         'INFO:%s:verbose' % logger.name,
+                                         'INFO:%s:notice' % logger.name,
+                                         'WARNING:%s:warning' % logger.name])
+
```
Can you add another test (or several) which makes malformed calls e.g.
- Too many arguments
- Too few arguments (or no arguments)
- Message is not a string
- Level is not redis.LOG_*

bmerry commented on fakeredis/_server.py:
```diff
@@ -13,12 +14,21 @@
 from collections import defaultdict
 from collections.abc import MutableMapping
 
+import lupa
```
Please remove this and import lupa only where it's needed. It's an optional dependency and fakeredis should work without it as long as Lua functionality isn't needed.

upcFrost replied:
Removed

upcFrost commented:
Added tests and fixed multiarg calls and type conversion. Redis actually does some strange things with types. String, int and float are all converted to string, while boolean is just ignored.

bmerry commented on fakeredis/_server.py:
```diff
@@ -23,10 +22,16 @@
+    1: logging.INFO,
+    2: logging.INFO,
+    3: logging.WARNING
 }
```
Is the double indirection because redis.LOG_DEBUG has the value 0 in real redis?

upcFrost replied:
Yes, and because user can actually use numbers instead of `redis.LOG_*` variables, just like in `logging`, but level values are different

bmerry commented on fakeredis/_server.py:
```diff
@@ -2324,13 +2332,18 @@ def _lua_redis_pcall(self, lua_runtime, expected_globals, op, *args):
+        msg = ' '.join([x.decode('utf-8')
+                        if isinstance(x, bytes) else str(x)
+                        for x in args if not isinstance(x, bool)])
+        LOGGER.log(REDIS_LOG_LEVELS_TO_LOGGING.get(lvl, logging.WARNING), msg)
```
Is it necessary to use `get` rather than just `[]`, given that you've already checked that `lvl` is in `REDIS_LOG_LEVELS.values()`?

bmerry commented on test_fakeredis.py:
```diff
@@ -4178,6 +4178,45 @@ def test_lua_log(self):
 
+    def test_lua_log_no_message(self):
+        if not isinstance(self.redis, (fakeredis.FakeRedis, fakeredis.FakeStrictRedis)):
+            pytest.skip("Works only on fakeredis, as testing real redis will require access to its logs")
```
Is this skip necessary? Won't real redis raise the same ResponseError? Same applies to the other tests that expect a ResponseError.

bmerry commented on test_fakeredis.py:
```diff
@@ -4178,6 +4178,45 @@ def test_lua_log(self):
+
+    def test_lua_log_defined_vars(self):
+        if not isinstance(self.redis, (fakeredis.FakeRedis, fakeredis.FakeStrictRedis)):
+            pytest.skip("Works only on fakeredis, as testing real redis will require access to its logs")
```
Since this code gets copy-pasted, it would be nice if it could be wrapped up into a decorator (so that it's stylistically similar to `@redis3_only` for example).
